cpp/source/basic.rst

```diff
@@ -48,3 +48,29 @@ boilerplate for you.  It will run the contained expression and check the resulti
 .. recipe:: ../code/basic_arrow.cc ReturnNotOk
   :caption: Using ARROW_RETURN_NOT_OK to check the status
   :dedent: 2
+
+
+Using the Visitor Pattern
+=========================
+
+Arrow classes like DataType, Scalar, and Array have specialized subclasses for
+each Arrow type. In order to work with them, use the visitor pattern. These 
+types provide an Accept method and have associated Visitor classes.
+
+As an example, below we implement a :cpp:class:`arrow::TypeVisitor` that counts
+the number of  primitive and nested types.
+
+To implement a TypeVisitor we have to implement a Visit method for every possible
+DataType we wish to handle. Fortunately, we can often use templates and type
+traits to make this less verbose.
+
+.. literalinclude:: ../code/basic_arrow.cc
+   :language: cpp
+   :linenos:
+   :start-at: class TypeCountVisitor
+   :end-at: };  // TypeCountVisitor
+   :caption: TypeVisitor that counts nested and non-nested types
```

Comment:
This is not a terrific example, because you can simply use the `is_nested` function in `arrow/type_traits.h` without having to use a visitor.

AFAIU, the cookbook tries to showcase how to achieve real-world-ish goals, not how to use a given API.
@westonpace @amol-  What do you think?

Reply:
Agreed. I'd like to find a better example, but haven't yet thought of one that wasn't too long. But as a first step I'd like to at least figure out how to use our API.

Reply:
IMO https://github.com/apache/arrow/pull/11886 was a really good example of a `TypeVisitor` :-)

Reply:
I've created two new examples. One is a class that generates a random record batch given a schema and row count, which showcases an approach to handling nested types.

The other is a class that can sum across columns of varied number types, which showcases how to use the type traits to avoid repetitive implementations for similar types.

I think these are better use cases that my original. The first is something I've seen in the wild (and is also inspired by our internal test functions).